state-synchronizer/src/coordinator.rs

```diff
@@ -102,31 +102,58 @@ impl PendingLedgerInfos {
         }
     }
 
+    // Update self based on a locally synced state:
+    // * remove outdated LIs
+    // * update target_li
     fn update(&mut self, sync_state: &SynchronizerState, chunk_limit: u64) {
+        // prune any pending LIs that were successfully committed
         let highest_committed_li = sync_state.highest_local_li.ledger_info().version();
+        self.pending_li_queue = self.pending_li_queue.split_off(&(highest_committed_li + 1));
+
+        // pick target LI to use for sending ProgressiveTargetType requests
         let highest_synced = sync_state.highest_version_in_local_storage();
+        self.target_li = self.get_next_target_li(highest_synced, chunk_limit);
+    }
 
-        // prune any pending LIs that were successfully committed
-        self.pending_li_queue = self.pending_li_queue.split_off(&(highest_committed_li + 1));
+    // Chooses a target li for a chunk request, under the (possibly optimistic) assumption that
+    // the locally synced version is (or will be) `sync_version`
+    fn get_next_target_li(
+        &self,
+        sync_version: u64,
+        chunk_limit: u64,
+    ) -> Option<LedgerInfoWithSignatures> {
+        let target_version = self
+            .target_li
+            .as_ref()
+            .map_or(0, |li| li.ledger_info().version());
+        if sync_version == target_version {
+            // current target version is satisfied, so pick the next target
 
-        // pick target LI to use for sending ProgressiveTargetType requests.
-        self.target_li = if highest_committed_li == highest_synced {
-            // try to find LI with max version that will fit in a single chunk
-            self.pending_li_queue
-                .range((Included(0), Included(highest_synced + chunk_limit)))
+            // first, try to find LI with max version that will fit in a single chunk
+            let li_within_single_chunk = self
+                .pending_li_queue
+                .range((Excluded(sync_version), Included(sync_version + chunk_limit)))
                 .rev()
                 .next()
-                .map(|(_version, ledger_info)| ledger_info.clone())
+                .map(|(_version, ledger_info)| ledger_info.clone());
+
+            if li_within_single_chunk.is_some() {
```

Comment:
you can use `or_else` here, but the logic in this function is so fancy. I'd rather just maintain a FIFO queue we request for.